File: backend/lib/cdp/config/elasticsearch.ex
defmodule Elasticsearch do
  require Lager

  def default_url do
    'http://localhost:9200'
  end

  def initialize do
    :inets.start
    {status, event} = :httpc.request(:put, {index_template_url, [], '', template_as_json}, [], [])
    case event do
      {{_, 200, _ },_, _ } -> status
      {{_, error_code, _ },_, _ }   ->
        log_error(event)
        error_code
      _ ->
        log_error(event)
        status
    end
  end

  defp log_error(event) do
    Lager.error("ES event indices mapping definition failed: #{inspect(event)}")
  end

  def index_template_url do
    :string.join([default_url, '_template', 'events_index_template'], '/')
  end

  defp template_as_json do
    JSEX.encode!([template: "#{index_prefix}*", mappings: [ event: [ properties: build_properties_mapping]]])
  end

  defp build_properties_mapping do
    map_fields Event.searchable_fields
  end

  defp map_fields(fields) do
    Enum.map fields, fn {name, type, properties} ->
      map_field(name, type, properties)
    end
  end

  defp map_field(field, :nested, nested_fields) do
    {field, [type: :nested, properties: map_fields(nested_fields)]}
  end

  defp map_field(field, type, _properties) do
    field_body = case type do
      :multi_field ->
            [
              fields: [
                {:analyzed, [{:type, :string}, {:index, :analyzed}]},
                {field, [{:type, :string}, {:index, :not_analyzed}]}
              ]
            ]
      _ -> [index: :not_analyzed]
    end
    {field, [type: type] ++ field_body}
  end

  def index_prefix do
    "cdp_institution_#{Mix.env}"
  end
end

File: config/initializers/elasticsearch.rb
require "cdx/api/elasticsearch"

module ElasticsearchHelper
  class << self
    def client
      Cdx::Api.client
    end

    def index_prefix
      "cdp_institution_#{Rails.env}"
    end

    def template
      {
        template: "#{index_prefix}*",
        mappings: {
          event: {
            properties: build_properties_mapping
          }
        }
      }
    end

    def build_properties_mapping
      map_fields Cdx::Api.searchable_fields
    end

    def map_fields fields
      fields.inject({}) do |properties, field|
        map_field properties, field
      end
    end

    def map_field properties, field
      properties[field[:name]] =
        if field[:type] == "nested"
          {type: :nested, properties: map_fields(field[:sub_fields])}
        else
          field_body =
            case field[:type]
            when "multi_field"
              {
                fields: {
                  analyzed: {type: :string, index: :analyzed},
                  field[:name] => {type: :string, index: :not_analyzed}
                }
              }
            else
              {index: :not_analyzed}
            end
          {type: field[:type]}.merge(field_body)
        end
      properties
    end
  end

  client.indices.put_template name: "events_index_template", body: template
end
